crates/torii/grpc/src/client.rs

```diff
@@ -40,18 +62,50 @@ pub struct WorldClient {
 
 impl WorldClient {
     #[cfg(not(target_arch = "wasm32"))]
-    pub async fn new<D>(dst: D, _world_address: Felt) -> Result<Self, Error>
-    where
-        D: TryInto<tonic::transport::Endpoint>,
-        D::Error: Into<Box<(dyn std::error::Error + Send + Sync + 'static)>>,
-    {
-        Ok(Self {
-            _world_address,
-            inner: world_client::WorldClient::connect(dst).await.map_err(Error::Transport)?,
-        })
+    pub async fn new(dst: String, _world_address: Felt) -> Result<Self, Error> {
+        let endpoint = Endpoint::from_shared(dst).map_err(|e| Error::Endpoint(e.to_string()))?;
+        let uri = endpoint.uri().clone();
+
+        // Localhost
+        if uri.to_string().starts_with("http://") {
+            return Ok(Self {
+                _world_address,
+                inner: world_client::WorldClient::connect(endpoint)
+                    .await
+                    .map_err(Error::Transport)?,
+            });
+        }
+
+        let host = uri.host().ok_or_else(|| Error::InvalidUri("Missing host".into()))?;
+        let connector = SslConnector::builder(SslMethod::tls()).map_err(Error::Ssl)?.build();
+        let tcp_stream = TcpStream::connect(&(host.to_owned() + ":443")).map_err(Error::Tcp)?;
+        let mut stream = connector.connect(&host, tcp_stream).map_err(Error::Tls)?;
+
+        stream.write_all(b"GET / HTTP/1.0\r\n\r\n").map_err(Error::Io)?;
+        let mut response = vec![];
+        stream.read_to_end(&mut response).map_err(Error::Io)?;
+
+        let certs = stream.ssl().peer_cert_chain().ok_or_else(|| Error::Ssl(ErrorStack::get()))?;
+        let mut certs_str = String::new();
+        for cert in certs {
+            let pem = cert.to_pem().map_err(Error::Ssl)?;
+            certs_str.push_str(std::str::from_utf8(&pem).map_err(Error::Utf8)?);
+        }
+
+        let config = ClientTlsConfig::new()
+            .ca_certificate(Certificate::from_pem(&certs_str))
+            .domain_name(host);
+
+        let channel = endpoint
+            .tls_config(config)
+            .map_err(|e| Error::TlsConfig(e.to_string()))?
+            .connect()
+            .await
+            .map_err(Error::Transport)?;
```

Comment:
Would be great to have the channel initialized with certificate `tls` in a function. 👍